File: python-algorithm/algo_09_math/J62.cpp
//
// Created by LiHeng on 2020/7/1.
//
#include <iostream>
using  namespace std;
class Solution {
public:
    int lastRemaining(int n, int m) {
        int p=0;
        for(int i = 2; i <= n; i++){
            p = (p+m)%i;
        }
        return p;
    }
};

//递推公式f(n,m) = [f(n-1,m)+m]%n

File: python-algorithm/algo_09_math/J62.py
# -*- coding: utf-8 -*-
# @Time    : 2020/10/6 22:59
# @Author  : No Name

class Solution:
    def lastRemaining(self, n: int, m: int) -> int:
        record = [i for i in range(n)]
        idx = 0
        while len(record) > 1:
            idx = (idx+m-1)%len(record)
            record.pop(idx)
        return record[0]


# 快多了，但是没想明白怎么推出来的公式
class Solution:
    def lastRemaining(self, n: int, m: int) -> int:
        p=0
        for i in range(2, n+1):
            p = (p+m)%i
        return p